File: api/schemas/create_domain/create_domain.py
import graphene

from graphql import GraphQLError

from app import logger
from db import db_session
from functions.auth_wrappers import require_token
from functions.auth_functions import is_user_write
from functions.input_validators import cleanse_input, cleanse_input_list
from models import (
    Organizations,
    Domains,
)
from scalars.url import URL
from scalars.slug import Slug
from scalars.selectors import Selectors


class CreateDomainInput(graphene.InputObjectType):
    """
    Input object used to define the argument fields for the createDomain
    mutation
    """

    org_slug = Slug(
        description="Organizations slug that you would like to connect "
        "this domain to.",
        required=True,
    )
    url = URL(
        description="URL that you would like to be added to database.", required=True,
    )
    selectors = Selectors(
        description="DKIM selector strings corresponding to this domain",
        required=False,
    )


class CreateDomain(graphene.Mutation):
    """
    Mutation used to create a new domain for an organization
    """

    class Arguments:
        input = CreateDomainInput(
            required=True, description="Input object containing all arguement fields."
        )

    status = graphene.Boolean()

    @require_token
    def mutate(self, info, **kwargs):
        user_id = kwargs.get("user_id")
        user_roles = kwargs.get("user_roles")
        org_slug = cleanse_input(kwargs.get("input", {}).get("org_slug"))
        domain = cleanse_input(kwargs.get("input", {}).get("url"))
        selectors = cleanse_input_list(kwargs.get("input", {}).get("selectors", []))

        # Check to see if org acronym is SA Org
        if org_slug == "super-admin":
            logger.warning(
                f"User: {user_id} tried to create a domain in the super admin org."
            )
            raise GraphQLError("Error, unable to create domain.")

        # Check to see if org exists
        org_orm = (
            db_session.query(Organizations)
            .filter(Organizations.slug == org_slug)
            .first()
        )

        if org_orm is None:
            logger.warning(
                f"User: {user_id} tried to create a domain in {org_slug}, but organization does not exist."
            )
            raise GraphQLError("Error, unable to create domain.")
        org_id = org_orm.id

        # Check to see if domain exists
        domain_orm = db_session.query(Domains).filter(Domains.domain == domain).first()

        if domain_orm is not None:
            logger.warning(
                f"User: {user_id} tried to create a domain that is already in use."
            )
            raise GraphQLError("Error, unable to create domain.")

        if is_user_write(user_roles=user_roles, org_id=org_id):
            new_domain = Domains(
                domain=domain, selectors=selectors, organization_id=org_id
            )
            try:
                db_session.add(new_domain)
                db_session.commit()
                logger.info(
                    f"User: {user_id} successfully created {domain} for the {org_slug} organization."
                )
                return CreateDomain(status=True)

            except Exception as e:
                db_session.rollback()
                db_session.flush()
                logger.error(
                    f"Database error occurred when user: {user_id} tried to create a domain for {org_slug} error: {str(e)}"
                )
                raise GraphQLError("Error, unable to create domain.")

        else:
            logger.warning(
                f"User: {user_id} tried to create a domain for this org: {org_slug}, but does not have the proper permissions."
            )
            raise GraphQLError("Error, unable to create domain.")

File: api-js/src/mutations/domain/create-domain.js
const { GraphQLNonNull, GraphQLList, GraphQLID } = require('graphql')
const { mutationWithClientMutationId, fromGlobalId } = require('graphql-relay')
const { t } = require('@lingui/macro')
const { Domain, Selectors } = require('../../scalars')
const { domainType } = require('../../types')

const createDomain = new mutationWithClientMutationId({
  name: 'CreateDomain',
  description: 'Mutation used to create a new domain for an organization.',
  inputFields: () => ({
    orgId: {
      type: GraphQLNonNull(GraphQLID),
      description:
        'The global id of the organization you wish to assign this domain to.',
    },
    domain: {
      type: GraphQLNonNull(Domain),
      description: 'Url that you would like to be added to the database.',
    },
    selectors: {
      type: new GraphQLList(Selectors),
      description: 'DKIM selector strings corresponding to this domain.',
    },
  }),
  outputFields: () => ({
    domain: {
      type: domainType,
      description: 'The newly created domain.',
      resolve: async (payload) => {
        return payload.domain
      },
    },
  }),
  mutateAndGetPayload: async (
    args,
    {
      i18n,
      request,
      query,
      collections,
      transaction,
      userKey,
      auth: { checkPermission, userRequired },
      loaders: { domainLoaderByDomain, orgLoaderByKey },
      validators: { cleanseInput },
    },
  ) => {
    // Cleanse input
    const { type: _orgType, id: orgId } = fromGlobalId(cleanseInput(args.orgId))
    const domain = cleanseInput(args.domain)

    let selectors
    if (typeof args.selectors !== 'undefined') {
      selectors = args.selectors.map((selector) => cleanseInput(selector))
    } else {
      selectors = []
    }

    // Get User
    await userRequired()

    // Check to see if org exists
    const org = await orgLoaderByKey.load(orgId)

    if (typeof org === 'undefined') {
      console.warn(
        `User: ${userKey} attempted to create a domain to an organization: ${orgId} that does not exist.`,
      )
      throw new Error(i18n._(t`Unable to create domain. Please try again.`))
    }

    // Check to see if user belongs to org
    const permission = await checkPermission({ orgId: org._id })

    if (
      permission !== 'user' &&
      permission !== 'admin' &&
      permission !== 'super_admin'
    ) {
      console.warn(
        `User: ${userKey} attempted to create a domain in: ${org.slug}, however they do not have permission to do so.`,
      )
      throw new Error(i18n._(t`Unable to create domain. Please try again.`))
    }

    const insertDomain = {
      domain: domain,
      lastRan: null,
      selectors: selectors,
    }

    // Check to see if domain already belongs to same org
    let checkDomainCursor
    try {
      checkDomainCursor = await query`
        LET domainIds = (FOR domain IN domains FILTER domain.domain == ${insertDomain.domain} RETURN { id: domain._id })
        FOR domainId IN domainIds 
          LET domainEdges = (FOR v, e IN 1..1 ANY domainId.id claims RETURN { _from: e._from })
            FOR domainEdge IN domainEdges
              LET org = DOCUMENT(domainEdge._from)
              FILTER org._key == ${org._key}
              RETURN MERGE({ _id: org._id, _key: org._key, _rev: org._rev }, TRANSLATE(${request.language}, org.orgDetails))
      `
    } catch (err) {
      console.error(
        `Database error occurred while running check to see if domain already exists in an org: ${err}`,
      )
      throw new Error(i18n._(t`Unable to create domain. Please try again.`))
    }

    const checkOrgDomain = await checkDomainCursor.next()

    if (typeof checkOrgDomain !== 'undefined') {
      console.warn(
        `User: ${userKey} attempted to create a domain for: ${org.slug}, however that org already has that domain claimed.`,
      )
      throw new Error(i18n._(t`Unable to create domain. Please try again.`))
    }

    // Check to see if domain already exists in db
    const checkDomain = await domainLoaderByDomain.load(insertDomain.domain)

    // Generate list of collections names
    const collectionStrings = []
    for (const property in collections) {
      collectionStrings.push(property.toString())
    }

    // Setup Transaction
    const trx = await transaction(collectionStrings)

    if (typeof checkDomain === 'undefined') {
      const insertedDomain = await trx.run(() =>
        collections.domains.save(insertDomain),
      )
      await trx.run(() =>
        collections.claims.save({ _from: org._id, _to: insertedDomain._id }),
      )
    } else {
      const selectorList = checkDomain.selectors
      selectors.forEach((selector) => {
        if (!checkDomain.selectors.includes(selector)) {
          selectorList.push(selector)
        }
      })
      insertDomain.selectors = selectorList

      await trx.run(
        async () =>
          await query`
        UPSERT { _key: ${checkDomain._key} }
          INSERT ${insertDomain}
          UPDATE ${insertDomain}
          IN domains
      `,
      )
      await trx.run(() =>
        collections.claims.save({ _from: org._id, _to: checkDomain._id }),
      )
    }

    try {
      await trx.commit()
    } catch (err) {
      console.error(
        `Database error occurred while committing create domain transaction: ${err}`,
      )
      throw new Error(i18n._(t`Unable to create domain. Please try again.`))
    }

    // Clear dataloader incase anything was updated or inserted into domain
    await domainLoaderByDomain.clear(insertDomain.domain)
    const returnDomain = await domainLoaderByDomain.load(insertDomain.domain)

    console.info(
      `User: ${userKey} successfully created ${returnDomain.domain} in org: ${org.slug}.`,
    )

    returnDomain.id = returnDomain._key
    return {
      domain: {
        ...returnDomain,
      },
    }
  },
})

module.exports = {
  createDomain,
}
